### PR TITLE
Store client_{id,secret} in token file #3477

### LINKED ISSUE #3477: mutt_oauth2.py to store client_id and client secret
## Expected Behaviour

When an access token expired, `mutt_oauth2.py` should use the refresh token to obtain a new access token.

## Actual Behaviour

`mutt_oauth2.py` fails because it does not submit the client ID and secret.

## Steps to Reproduce

1. Do not modify `mutt_oauth2.py`,
2. follow the instructions to register a new client (I used Google, but it should not matter),
3. `mutt_oauth2.py --authorize` to obtain an initial set of tokens, eg:
```
mutt_oauth2.py \
	--verbose \
	--debug \
        --provider google \
	--authorize \
	--client-id "${client_id}" \
	--client-secret "${client_secret}" \
        --authflow authcode \
	--encryption-pipe "gpg --encrypt --recipient frank.gruellich@gmail.com" \
	~/.neomutt/tokens/frank.gruellich@gmail.com.gpg
```
4. wait until the access token expire (or modify the code to always flag them as expired),
5. `mutt_oauth2.py --test` the connection results in:
```
...
400 Bad Request
invalid_request
Could not determine client ID from request.
Perhaps refresh token invalid. Try running once with "--authorize"
```

Closer inspections reveal the it submit an empty client_id and client_secret.

## How often does this happen?

Always, whenever the token expires.

## When did it start to happen?
I think when the idea surfaced that `mutt_oauth2.py` should be usable without any modifications, eg. by coding client ID and secrets into it.

## NeoMutt Version
Should matter, NeoMutt was installed from homebrew, `mutt_oauth2.py` was taken from current `main` branch.
```
NeoMutt 20220429
Copyright (C) 1996-2022 Michael R. Elkins and others.
NeoMutt comes with ABSOLUTELY NO WARRANTY; for details type 'neomutt -vv'.
NeoMutt is free software, and you are welcome to redistribute it
under certain conditions; type 'neomutt -vv' for details.

System: Darwin 21.5.0 (arm64)
ncurses: ncurses 6.3.20211021 (compiled with 6.3.20211021)
libiconv: 1.11
libidn2: 2.3.2 (compiled with 2.3.2)
GPGME: 1.17.1
OpenSSL: OpenSSL 1.1.1n  15 Mar 2022
libnotmuch: 5.6.0
storage: tokyocabinet, lmdb

Configure options: --prefix=/opt/homebrew/Cellar/neomutt/20220429 --gss --disable-idn --idn2 --lmdb --notmuch --sasl --tokyocabinet --with-gpgme=/opt/homebrew/opt/gpgme --with-lua=/opt/homebrew/opt/lua --with-ssl=/opt/homebrew/opt/openssl@1.1 --with-ui=ncurses

Compilation CFLAGS: -g -O2 -std=c99 -D_ALL_SOURCE=1 -D_GNU_SOURCE=1 -D__EXTENSIONS__ -I/opt/homebrew/opt/gpgme/include -I/opt/homebrew/Cellar/libassuan/2.5.5/include -I/opt/homebrew/opt/libgpg-error/include -I/opt/homebrew/Cellar/neomutt/20220429/include -I/opt/homebrew/opt/lua/include/lua5.4 -DNCURSES_WIDECHAR -I/opt/homebrew/opt/openssl@1.1/include

Default options:
  +attach_headers_color +compose_to_sender +compress +cond_date +debug 
  +encrypt_to_self +forgotten_attachments +forwref +ifdef +imap +index_color 
  +initials +limit_current_thread +multiple_fcc +nested_if +new_mail +nntp +pop 
  +progress +quasi_delete +regcomp +reply_with_xorig +sensible_browser +sidebar 
  +skip_quoted +smtp +status_color +timeout +tls_sni +trash 

Compile options:
  -autocrypt +fcntl -flock -fmemopen +futimens +getaddrinfo -gnutls +gpgme +gss 
  +hcache -homespool +idn -inotify -locales_hack +lua -mixmaster +nls +notmuch 
  +openssl +pgp +regex +sasl +smime -sqlite +sun_attachment 

MAILPATH="/var/mail"
PKGDATADIR="/opt/homebrew/Cellar/neomutt/20220429/share/neomutt"
SENDMAIL="/usr/sbin/sendmail"
SYSCONFDIR="/opt/homebrew/Cellar/neomutt/20220429/etc"

To learn more about NeoMutt, visit: https://neomutt.org
If you find a bug in NeoMutt, please raise an issue at:
    https://github.com/neomutt/neomutt/issues
or send an email to: <neomutt-devel@neomutt.org>
```

## Extra Info

No relevant information...

### PR DESCRIPTION
**Pull request by FrankGruellich**
* **What does this PR do?**
Makes `mutt_oauth2.py` store / fetch client ID and secret into / from token file.

* **Screenshots (if relevant)**

* **Does this PR meet the acceptance criteria?** (This is just a reminder for you,
  this section can be removed if you fulfill it.)

   - Documentation created/updated (you have to edit
     [docs/manual.xml.head](https://www.github.com/neomutt/neomutt/blob/main/docs/manual.xml.head)
     for that)

   - All builds and tests are passing

   - Added [doxygen code documentation](https://neomutt.org/dev/doxygen)
     [syntax](http://www.stack.nl/~dimitri/doxygen/manual/docblocks.html)

   - Code follows the [style guide](https://neomutt.org/dev/code)

* **What are the relevant issue numbers?**
This tries to fix #3477

### CODE REVIEW

**Comment by flatcap** [comment]
Thanks for the improvements @FrankGruellich :heart:

@holmanb @brunnre8 @ryanakca 
As people who understand this better than I...
any thoughts or suggestions?

Thanks

**Comment by ryanakca** [comment]
There's something slightly buggy in the resulting script. The `--provider X` option is not required, and the script interactively prompts users for a provider during authorization if they did not pass a `--provider` option. In this case, however, authorization fails because the script does not send a `client_id`. For example, attempting to get a token from Microsoft using
```
echo -e 'microsoft\nlocalhostauthcode\nneomutt@mcgill.ca\n' | \
   ~/bin/mutt_oauth2.py --authorize \
                        --client-id='08162f7c-0fd2-4200-a84a-f25a4db0b584' \
                        --client-secret='TxRBilcHdC6WGBee]fs?QR:SJ8nI[g82' \
                        --encryption-pipe "cat" \
                        --decryption-pipe "cat" ~/.mutt/xoauth2-mcgill.ca-token
```
results in `AADSTS900144: The request body must contain the following parameter: 'client_id'.` (This does not happen if one passes `--provider microsoft` along with `--authorize`.) 

I suspect the culprit involves lines 115-118 and lines 160-161 (but I haven't carefully traced this through). Lines 115-118 update `registrations[P]['client_id']` (and the corresponding client_secret entry) if users pass `--provider P`, but lines 160-161 do not similarly update `registrations[P]` if they interactively get the provider. Perhaps these lines should be refactored so that we're not storing and fetching client_ids and client_secrets to/from multiple places?

**Comment by the-x-at** [comment]
As far as I understand the script, the original author intended `mutt_oauth2.py` to be taken and modified, as `client_secret` needs to be inserted into the source code. The same for the key store pipeline, where `YOUR_GPG_IDENTITY` needs to be modified in the source. This connects to this issue, and I would suggest to update the command line interface of `mutt_oauth2.py` such that no script modification is required anymore and rethink what should go into the token store, see also issue #3442 where we would like to enhance the documentation for this authentication method.

**Comment by holmanb** [comment]
> As far as I understand the script, the original author intended mutt_oauth2.py to be taken and modified, as client_secret needs to be inserted into the source code. The same for the key store pipeline, where YOUR_GPG_IDENTITY needs to be modified in the source. This connects to this issue, and I would suggest to update the command line interface of mutt_oauth2.py such that no script modification is required anymore and 

@the-x-at for context on recent changes in script behavior: https://github.com/neomutt/neomutt/commit/7decc96c7ccabae4cb940927665783fe6ba81761

**Comment by the-x-at** [comment]
Hi, @holmanb, thanks for pointing me to this. I need to update the repository then. That way I can provide more useful comments in the future. Sorry for that.

**Comment by FrankGruellich** [comment]
> There's something slightly buggy in the resulting script. The `--provider X` option is not required, and the script interactively prompts users for a provider during authorization if they did not pass a `--provider` option. [great description]

Indeed, in that situation the script still tried to fetch information from the `registrations`, sorry for missing that.  I've fixed that, your test case does not throw an error right away and should succeed now.  Please try again and let me know.

File: contrib/oauth2/mutt_oauth2.py [hunk 1]
```diff
@@ -167,7 +170,7 @@ def writetokenfile():
 if args.authflow:
     authflow = args.authflow
 
-baseparams = {'client_id': registration['client_id']}
+baseparams = {'client_id': args.client_id or registration['client_id']}
```
Comment (brunnre8) [inline contrib/oauth2/mutt_oauth2.py:173]: Getting it from registration is pointless.
Unless the user modified the script by hardcoding those values, it doesn't do anything and if they did they wouldn't pull in your changes.

If we want to swtich to a version where the users doesn't monkeypatch the script file prior to running it we can simply remove that case.

File: contrib/oauth2/mutt_oauth2.py [hunk 2]
```diff
@@ -260,7 +263,7 @@ def do_GET(self):
             del p[k]
         p.update({'grant_type': 'authorization_code',
                   'code': authcode,
-                  'client_secret': registration['client_secret'],
+                  'client_secret': args.client_secret or registration['client_secret'],
```
Comment (brunnre8) [inline contrib/oauth2/mutt_oauth2.py:266]: same

File: contrib/oauth2/mutt_oauth2.py (hunk 1 above)
Comment (FrankGruellich) [reply contrib/oauth2/mutt_oauth2.py:173]: > Getting it from registration is pointless. Unless the user modified the script by hardcoding those values, it doesn't do anything and if they did they wouldn't pull in your changes.

Understood.  But I tried to only add functionality and not remove any, so I left this there for compatibility reasons, if users prefer to edit their secrets into the script (for whatever reasons).  What do you think?  I'm not going to argue much, though.

> If we want to swtich to a version where the users doesn't monkeypatch the script file prior to running it we can simply remove that case.

Honestly, I don't think it makes much sense to modify the current script into that direction.  I don't want to offend anyone, but script is quite messy and IMHO difficult to modify.  I don't know if I could do any better, but I could give it a try.

Comment (holmanb) [reply contrib/oauth2/mutt_oauth2.py:173]: > > Getting it from registration is pointless. Unless the user modified the script by hardcoding those values, it doesn't do anything and if they did they wouldn't pull in your changes.
> 
> Understood.  But I tried to only add functionality and not remove any, so I left this there for compatibility reasons, if users prefer to edit their secrets into the script (for whatever reasons).  What do you think?  I'm not going to argue much, though.

I believe backwards compatibility is not required here. Backwards compatibility for editing your config into the script was broken recently, in favor of cli args that can go directly into your invokation in your .muttrc

Comment (FrankGruellich) [reply contrib/oauth2/mutt_oauth2.py]: > I believe keeping backwards compatibility is not required.

Okay, fine with me.  Fixed!

File: contrib/oauth2/mutt_oauth2.py (hunk 2 above)
Comment (FrankGruellich) [reply contrib/oauth2/mutt_oauth2.py]: > same

Fixed!

File: contrib/oauth2/mutt_oauth2.py [hunk 3]
```diff
@@ -149,13 +149,16 @@ def writetokenfile():
     if not args.authorize:
         sys.exit('You must run script with "--authorize" at least once.')
     print('Available app and endpoint registrations:', *registrations)
```
Comment (holmanb) [inline contrib/oauth2/mutt_oauth2.py:151]: If the provider is passed as an argument, this print is not useful. Can we integrate this into the input() call for registration?

File: contrib/oauth2/mutt_oauth2.py [hunk 4]
```diff
@@ -149,13 +149,16 @@ def writetokenfile():
...
-                              'or "devicecode"): ')
+    token['registration'] = args.provider or input('OAuth2 registration: ')
+    token['authflow'] = args.authflow or input(
+        'Preferred OAuth2 flow ("authcode" or "localhostauthcode" or "devicecode"): '
+    )
     token['email'] = input('Account e-mail address: ')
     token['access_token'] = ''
     token['access_token_expiration'] = ''
     token['refresh_token'] = ''
+    token['client_id'] = args.client_id
```
Comment (holmanb) [inline contrib/oauth2/mutt_oauth2.py:160]: Could we make client_id and client_secret also optionally interactive? If not they are required args but the script will not warn that there is a problem if they aren't included but will create a bogus url - not a great user experience.

File: contrib/oauth2/mutt_oauth2.py [hunk 5]
```diff
@@ -149,13 +149,16 @@ def writetokenfile():
...
         sys.exit('You must run script with "--authorize" at least once.')
     print('Available app and endpoint registrations:', *registrations)
-    token['registration'] = input('OAuth2 registration: ')
-    token['authflow'] = input('Preferred OAuth2 flow ("authcode" or "localhostauthcode" '
-                              'or "devicecode"): ')
+    token['registration'] = args.provider or input('OAuth2 registration: ')
+    token['authflow'] = args.authflow or input(
+        'Preferred OAuth2 flow ("authcode" or "localhostauthcode" or "devicecode"): '
+    )
     token['email'] = input('Account e-mail address: ')
```
Comment (holmanb) [inline contrib/oauth2/mutt_oauth2.py:156]: If we add an email cli arg in addition to the client_id and client_secret input() below, then auth can be done completely non-interactively, which would be nice.

**Review by holmanb — CHANGES_REQUESTED** [review]
Hi @FrankGruellich,

I added a couple small change requests, but based on some testing and modifications I did in my own branch tonight I think this is looking pretty good. Let me know your thoughts.

My own changes are in a different PR, but I plan on dropping the functional changes from my PR once this lands (while keeping the docs changes).

Thanks for this contribution!

**Comment by FrankGruellich** [comment]
Hi @holmanb,

thanks a lot for the suggestions.  I've added them as good as possible.  Let me know what you think.

Kind regards,
 Frank.

**Review by holmanb — APPROVED** [review]
I tested this locally with a gmail account and the `authcode` option.

Looks good to me, thanks!

**Comment by holmanb** [comment]
@brunnre8 @ryanakca Any other concerns? I'm in favor of merging.

**Comment by brunnre8** [comment]
I do not have any objections, but I also don't use the script.

The script itself is a horrifying mess in my view, but that's no fault of the PR author.
It crams to many different things into one interface... Subcommands would help rather than a single thing as that would make the validation of the input args much easier.

Also I'm not exactly fond of interactive scripts, they should just error out when they don't receive what they need. Especially if it's something that's usually run in the background such as this.

But as said, no fault of the PR author and the changes look reasonable when just reading the code, I didn't runtime test it.

**Comment by ryanakca** [comment]
Tested token creation locally with a Microsoft account and it looks good to me. Perhaps mention in the release notes that users will need to delete their old token file after upgrading. Otherwise, the script aborts with:

```
Traceback (most recent call last):
  File "/tmp/neomutt/contrib/oauth2/./mutt_oauth2.py", line 167, in <module>
    baseparams = {'client_id': token['client_id']}
KeyError: 'client_id'
```

**Comment by FrankGruellich** [comment]
> The script itself is a horrifying mess in my view, [...]

I don't want to offend the original contributors, but I admit that the code is indeed quite messy.  For this simple change I had to touch too many areas scattered all other the place of it.  I am thinking about rewriting the whole thing.

> It crams to many different things into one interface... Subcommands would help rather than a single thing as that would make the validation of the input args much easier.
> 
> Also I'm not exactly fond of interactive scripts, they should just error out when they don't receive what they need. Especially if it's something that's usually run in the background such as this.

I generally agree to that and will try to keep these suggestions in mind.

> ```
> Traceback (most recent call last):
>   File "/tmp/neomutt/contrib/oauth2/./mutt_oauth2.py", line 167, in <module>
>     baseparams = {'client_id': token['client_id']}
> KeyError: 'client_id'
> ```

I could try to work around that error.  OTOH, I don't think it's worth the effort.

**Comment by brunnre8** [comment]
nothing specific. Just noticed that a lot of the complexity stems from the overload

**Comment by holmanb** [comment]
Docs coverage for changes here is reflected in #3483, which adds other docs improvements as well

**Review by flatcap — APPROVED** [review]

**Comment by flatcap** [comment]
Thanks everyone!  :heart:

> I don't want to offend the original contributors

It's fine, they're Mutt people, not NeoMutt people :-)

> I am thinking about rewriting the whole thing

That sounds great.
I've just started a Discussion #3505 -- all ideas welcome.

**Pull request merged**